app/services/schedule_data_manager_service.rb

```diff
@@ -17,29 +18,63 @@ def self.calendar_cache_key(calendar)
   end
 
   def holidays
-    @holidays ||= Rails.cache.fetch(self.class.calendar_cache_key(@calendar), expires_in: 3.hours) do
-      hol_data = HolidayData.new
+    @holidays ||= find_holidays
+  end
 
-      Calendar.find(@calendar.id).outside_term_time.order(:start_date).includes(:academic_year, :calendar_event_type).map do |holiday|
-        academic_year = nil # Not really being used at the moment by the analytics code
+  def temperatures
+    @temperatures ||= find_temperatures
+  end
 
-        analytics_holiday = Holiday.new(
-          holiday.calendar_event_type.analytics_event_type.to_sym,
-          nil,
-          holiday.start_date,
-          holiday.end_date,
-          academic_year
-        )
+  def solar_pv
+    @solar_pv ||= find_solar_pv
+  end
 
-        hol_data.add(analytics_holiday)
+  def uk_grid_carbon_intensity
+    @uk_grid_carbon_intensity ||= find_uk_grid_carbon_intensity
+  end
+
+  private
+
+  def school_reading_date_bounds_present?
+    school_reading_date_bounds.present? && school_reading_date_bounds.uniq.size == 2
+  end
+
+  def find_solar_pv
+    cache_key = "#{@solar_pv_tuos_area_id}-solar-pv-2-tuos"
+    cached_solar_pv ||= Rails.cache.fetch(cache_key, expires_in: 3.hours) do
+      data = SolarPV.new('solar pv')
+      DataFeeds::SolarPvTuosReading.where(area_id: @solar_pv_tuos_area_id).pluck(:reading_date, :generation_mw_x48).each do |date, values|
+        data.add(date, values.map(&:to_f))
       end
-      Holidays.new(hol_data)
+      data
     end
+
+    # Only use solar pv data within lower datetime bounds of school meter readings
+    return cached_solar_pv unless @school.minimum_reading_date.present?
+
+    dates_to_remove = cached_solar_pv.keys.select { |date| date < @school.minimum_reading_date }
+    cached_solar_pv.remove_dates!(*dates_to_remove)
   end
 
-  def temperatures
+  def find_uk_grid_carbon_intensity
+    cache_key = "co2-feed"
+    cached_uk_grid_carbon_intensity ||= Rails.cache.fetch(cache_key, expires_in: 3.hours) do
+      uk_grid_carbon_intensity_data = GridCarbonIntensity.new
+      DataFeeds::CarbonIntensityReading.all.pluck(:reading_date, :carbon_intensity_x48).each do |date, values|
+        uk_grid_carbon_intensity_data.add(date, values.map(&:to_f))
+      end
+      uk_grid_carbon_intensity_data
+    end
+
+    # Only use uk grid carbon intensity data within lower datetime bounds of school meter readings
+    return cached_uk_grid_carbon_intensity unless @school.minimum_reading_date.present?
+
+    cached_uk_grid_carbon_intensity.select { |date, _v| date >= @school.minimum_reading_date }
```

Comment:
This will return a new Hash I believe, rather than an instance of GridCarbonIntensity. The analytics will be expecting an instance of the latter, so I think this logic should be similar to the other methods where its deleting entries from the object?